File: java/socketd/src/main/java/org/noear/socketd/transport/core/SessionBase.java
package org.noear.socketd.transport.core;

import java.util.HashMap;
import java.util.Map;
import java.util.Objects;

/**
 * 会话基类
 *
 * @author noear
 */
public abstract class SessionBase implements Session {
    protected final Channel channel;

    private final String sessionId;

    public SessionBase(Channel channel) {
        this.channel = channel;
        this.sessionId = generateId();
    }

    /**
     * 会话的附件与通道的各自独立
     */
    private Map<String, Object> attrMap;

    @Override
    public Map<String, Object> attrMap() {
        if (attrMap == null) {
            attrMap = new HashMap<>();
        }

        return attrMap;
    }

    /**
     * 是有属性
     *
     * @param name 名字
     */
    @Override
    public boolean attrHas(String name) {
        return attrMap.containsKey(name);
    }

    /**
     * 获取附件
     */
    @Override
    public <T> T attr(String name) {
        if (attrMap == null) {
            return null;
        }

        return (T) attrMap.get(name);
    }

    /**
     * 获取属性或默认值
     *
     * @param name 名字
     * @param def  默认值
     */
    @Override
    public <T> T attrOrDefault(String name, T def) {
        T tmp = attr(name);
        if (tmp == null) {
            return def;
        } else {
            return tmp;
        }
    }

    /**
     * 设置附件
     */
    @Override
    public <T> void attr(String name, T value) {
        if (attrMap == null) {
            attrMap = new HashMap<>();
        }
        attrMap.put(name, value);
    }

    @Override
    public String sessionId() {
        return sessionId;
    }

    protected String generateId() {
        return channel.getConfig().getIdGenerator().generate();
    }

    //不要加 hashCode, equals 重写！
}

File: python/socketd/socketd/core/SessionBase.py
from abc import ABC
from typing import Any, Dict
from .Session import Session
from .Channel import Channel


class SessionBase(Session, ABC):
    def __init__(self, channel: Channel):
        self.channel = channel
        self._attr_map = None
        self._session_id = None

    def get_attr_map(self) -> Dict[str, Any]:
        if self._attr_map is None:
            self._attr_map = {}

        return self._attr_map

    def get_attr(self, name: str) -> Any:
        if self._attr_map is None:
            return None

        return self._attr_map.get(name)

    def get_attr_or_default(self, name: str, default: Any) -> Any:
        tmp = self.get_attr(name)
        if tmp is None:
            return default
        else:
            return tmp

    def set_attr(self, name: str, value: Any) -> None:
        if self._attr_map is None:
            self._attr_map = {}
        self._attr_map[name] = value

    def get_session_id(self) -> str:
        if self._session_id is None:
            self._session_id = self.generate_id()

        return self._session_id

    def generate_id(self) -> str:
        return self.channel.get_config().get_id_generator()().__str__()

    def set_session_id(self, value):
        self._session_id = value
